components/raftstore-v2/src/operation/query/mod.rs

```diff
@@ -8,28 +8,75 @@
 //! first. Lease maintainance is implemented in lease module.
 //!
 //! Status query is implemented in the root module directly.
+//! Follower's read index is implemenented follower_read_index
+//! Leader's read index is implemented in read_index
+//! stale read check is implemented replica_read.rs
 
 use engine_traits::{KvEngine, RaftEngine};
 use kvproto::raft_cmdpb::{RaftCmdRequest, RaftCmdResponse, StatusCmdType};
 use raftstore::{
-    store::{cmd_resp, region_meta::RegionMeta, util, GroupState, ReadCallback},
+    store::{
+        cmd_resp, region_meta::RegionMeta, util, util::LeaseState, GroupState, ReadCallback,
+        RequestInspector, RequestPolicy,
+    },
     Error, Result,
 };
 use tikv_util::box_err;
+use txn_types::WriteBatchFlags;
 
 use crate::{
     fsm::PeerFsmDelegate,
     raft::Peer,
-    router::{DebugInfoChannel, QueryResChannel, QueryResult},
+    router::{DebugInfoChannel, QueryResChannel, QueryResult, ReadResponse},
 };
 
+mod follower_read_index;
 mod local;
+mod read_index;
+mod replica_read;
+
+impl<'a, EK: KvEngine, ER: RaftEngine, T: raftstore::store::Transport>
+    PeerFsmDelegate<'a, EK, ER, T>
+{
+    fn inspect_read(&mut self, req: &RaftCmdRequest) -> Result<RequestPolicy> {
```

Comment:
Although we expect there should be only read requests, but we need to check if it's misused.

Reply:
All read related checks from `pre_propose_raft_command` should also be checked here.

Reply:
it's checked in validate_query_msg.